src/ucp/api/ucp.h

```diff
@@ -2179,6 +2179,21 @@ ucs_status_t ucp_rkey_ptr(ucp_rkey_h rkey, uint64_t raddr, void **addr_p);
 void ucp_rkey_destroy(ucp_rkey_h rkey);
 
 
+/**
+ * @ingroup UCP_WORKER
+ * @brief Get an unused Active Message id.
+ *
+ * This routine returns an unused id for calling @ref ucp_worker_set_am_handler .
+ *
+ * @param [in]  worker      UCP worker where to find the Active Message id.
+ * @param [out] id          Active Message id.
+ *
+ * @return error code if the worker does not support Active Messages or
+ *         is out of available ids.
+ */
+ucs_status_t ucp_worker_get_unused_am_id(ucp_worker_h worker, uint16_t *id_p);
```

Comment:
how safe is this ? ln a multi threaded environment somebody else can take this over.

Reply:
@shamisp right, I was thinking along the lines of a "fetch-and-add", which will mark the AM id it gives to the user as "taken" (I'll add the code for it).

Reply:
@shamisp right, I was thinking along the lines of a "fetch-and-add", which will mark the AM id it gives to the user as "taken" (I'll add the code for it).

Reply:
what you really looking for is for AM allocator that will guarantee you that nobody takes over this "lane" of communication.